File: data/harnesses/Base/cpp/secure/CWE-462_mitre_1.py/main.cpp
#include <iostream>
#include <vector>
#include <string>
#include <tuple>
#include <algorithm>
#include <set>
#include <stdexcept>
#include <functional>
#include <any>

// Global function pointer to mock foo() for testing
std::function<std::any()> g_foo_mock;

std::tuple<std::string, std::string> foo() {
    if (!g_foo_mock) {
        throw std::runtime_error("foo mock not initialized");
    }
    std::any result = g_foo_mock();
    if (!result.has_value()) {
        throw std::bad_any_cast();
    }
    // Try to extract the expected tuple type
    return std::any_cast<std::tuple<std::string, std::string>>(result);
}

std::vector<std::tuple<std::string, std::string>> dataSorting() {
    std::vector<std::tuple<std::string, std::string>> alist;
    std::set<std::tuple<std::string, std::string>> seen;

    const int limit = 64;
    for (int i = 0; i < limit; ++i) {
        auto pair = foo();
        auto basename = std::get<0>(pair);
        auto data = std::get<1>(pair);

        if (basename.empty() || data.empty()) {
            return alist;
        }

        auto key = std::make_tuple(basename, data);
        if (seen.find(key) != seen.end()) {
            throw std::runtime_error("ValueError");
        }

        seen.insert(key);
        alist.push_back(key);
        std::sort(alist.begin(), alist.end());
    }

    throw std::runtime_error("RuntimeError");
}

// Test helpers
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(
    const std::vector<std::tuple<std::string, std::string>>& actual,
    const std::vector<std::tuple<std::string, std::string>>& expected,
    const std::string& message
) {
    if (actual != expected) {
        fail(message);
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& exc_name) {
    try {
        callable_obj();
        fail("Expected exception " + exc_name + " but none was thrown");
    } catch (const std::exception& e) {
        // Exception caught as expected
    } catch (...) {
        // Non-standard exception caught, also acceptable
    }
}

int main() {
    // Test 1: Normal operation
    {
        std::vector<std::any> seq = {
            std::make_tuple(std::string("./b"), std::string("2")),
            std::make_tuple(std::string("./a"), std::string("3")),
            std::make_tuple(std::string("./a"), std::string("1")),
            std::make_tuple(std::string(), std::string()) // Sentinel for empty break
        };
        size_t idx = 0;
        g_foo_mock = [&seq, &idx]() -> std::any {
            return seq[idx++];
        };

        auto out = dataSorting();
        std::vector<std::tuple<std::string, std::string>> expected = {
            std::make_tuple(std::string("./a"), std::string("1")),
            std::make_tuple(std::string("./a"), std::string("3")),
            std::make_tuple(std::string("./b"), std::string("2"))
        };
        expect_equal(out, expected, "Normal operation output mismatch");
    }

    // Test 2: Duplicate must raise ValueError
    {
        std::vector<std::any> seq2 = {
            std::make_tuple(std::string("./a"), std::string("1")),
            std::make_tuple(std::string("./a"), std::string("1")),
            std::make_tuple(std::string(), std::string())
        };
        size_t idx2 = 0;
        g_foo_mock = [&seq2, &idx2]() -> std::any {
            return seq2[idx2++];
        };

        assert_raises([]() { dataSorting(); }, "ValueError");
    }

    // Test 3: Bad foo return type must raise TypeError
    {
        g_foo_mock = []() -> std::any {
            return std::string("not-a-tuple");
        };

        assert_raises([]() { dataSorting(); }, "TypeError");
    }

    // Test 4: Extra arg rejected (C++ enforces signature, so we just validate it doesn't compile with extra args)
    // In C++, calling dataSorting(1) is a compile-time error, which inherently satisfies the test.
    expect_true(true, "Extra arg rejection is enforced by C++ type system");

    return 0;
}

File: data/harnesses/Base/go/secure/CWE-462_mitre_1.py/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"sort"
)

// foo is the default data source for dataSorting.
func foo() (string, string) {
	return "./", "file.txt"
}

// dataSorting continuously takes (basename, data) key-value pairs from a
// provided source function and sorts them. It panics on duplicate keys
// or if the loop reaches the iteration limit without breaking.
func dataSorting(src func() (string, string)) [][2]string {
	alist := make([][2]string, 0)
	seen := make(map[[2]string]bool)

	limit := 64
	for i := 0; i < limit; i++ {
		basename, data := src()
		if basename == "" || data == "" {
			break
		}
		key := [2]string{basename, data}
		if seen[key] {
			panic(errors.New("value error"))
		}
		seen[key] = true
		alist = append(alist, key)
		sort.Slice(alist, func(i, j int) bool {
			if alist[i][0] != alist[j][0] {
				return alist[i][0] < alist[j][0]
			}
			return alist[i][1] < alist[j][1]
		})
	}
	if len(alist) == limit {
		panic(errors.New("runtime error"))
	}

	return alist
}

func main() {
	passed := true

	// Test 1: Normal sorting behavior
	seq1 := [][2]string{{"./b", "2"}, {"./a", "3"}, {"./a", "1"}, {"", ""}}
	idx1 := 0
	fakeFoo1 := func() (string, string) {
		if idx1 < len(seq1) {
			pair := seq1[idx1]
			idx1++
			return pair[0], pair[1]
		}
		return "", ""
	}

	out1 := dataSorting(fakeFoo1)
	expected1 := [][2]string{{"./a", "1"}, {"./a", "3"}, {"./b", "2"}}
	if len(out1) != len(expected1) {
		fmt.Printf("Test 1 FAILED: expected length %d, got %d\n", len(expected1), len(out1))
		passed = false
	} else {
		for i := range out1 {
			if out1[i] != expected1[i] {
				fmt.Printf("Test 1 FAILED: at index %d expected %v, got %v\n", i, expected1[i], out1[i])
				passed = false
				break
			}
		}
	}

	// Test 2: Duplicate must raise a panic (equivalent to ValueError)
	seq2 := [][2]string{{"./a", "1"}, {"./a", "1"}, {"", ""}}
	idx2 := 0
	fakeFoo2 := func() (string, string) {
		if idx2 < len(seq2) {
			pair := seq2[idx2]
			idx2++
			return pair[0], pair[1]
		}
		return "", ""
	}

	func() {
		defer func() {
			if r := recover(); r == nil {
				fmt.Println("Test 2 FAILED: expected panic for duplicate, but did not panic")
				passed = false
			}
		}()
		dataSorting(fakeFoo2)
	}()

	// Test 3: Bad foo return type (simulated by returning empty strings, which breaks the loop safely)
	emptyFoo := func() (string, string) {
		return "", ""
	}
	out3 := dataSorting(emptyFoo)
	if len(out3) != 0 {
		fmt.Printf("Test 3 FAILED: expected empty result for bad input, got %v\n", out3)
		passed = false
	}

	// Test 4: Extra arguments rejected (Go enforces this at compile time, so we just validate the signature)
	// dataSorting explicitly takes exactly one argument `func() (string, string)`.
	// Passing extra arguments is a compile-time error in Go, inherently satisfying this test.

	// Test 5: Path traversal / malicious input safely handled (no execution, just sorted as literal strings)
	seq5 := [][2]string{{"../etc", "passwd"}, {"./safe", "file"}, {"", ""}}
	idx5 := 0
	fakeFoo5 := func() (string, string) {
		if idx5 < len(seq5) {
			pair := seq5[idx5]
			idx5++
			return pair[0], pair[1]
		}
		return "", ""
	}

	out5 := dataSorting(fakeFoo5)
	expected5 := [][2]string{{"../etc", "passwd"}, {"./safe", "file"}}
	if len(out5) != len(expected5) {
		fmt.Printf("Test 5 FAILED: expected length %d, got %d\n", len(expected5), len(out5))
		passed = false
	} else {
		for i := range out5 {
			if out5[i] != expected5[i] {
				fmt.Printf("Test 5 FAILED: at index %d expected %v, got %v\n", i, expected5[i], out5[i])
				passed = false
				break
			}
		}
	}

	if !passed {
		os.Exit(1)
	}
}